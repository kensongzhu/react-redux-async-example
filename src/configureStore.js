import thunkMiddleware from "redux-thunk";
import {createLogger} from "redux-logger";
import {applyMiddleware, createStore} from "redux"

import rootReducer from "reducers"

const loggerMiddleware = createLogger();

export default function(preloadedState) {

    return createStore(
        rootReducer,
        preloadedState,
        applyMiddleware(thunkMiddleware, loggerMiddleware)
    )
}