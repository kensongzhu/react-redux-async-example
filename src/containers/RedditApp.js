import React from "react";
import {connect} from "react-redux";
import App from "../components/App";
import {fetchPostsIfNeeded, invalidateSubreddit, selectSubreddit} from "../actions";

class RedditApp extends React.Component {

	componentDidMount() {
		const {dispatch, selectedSubreddit} = this.props;
		dispatch(fetchPostsIfNeeded(selectedSubreddit));
	}

	componentDidUpdate(prevProps, prevState, snapshot) {
		if (prevProps.selectedSubreddit !== this.props.selectedSubreddit) {
			const {dispatch, selectedSubreddit} = this.props;
			dispatch(fetchPostsIfNeeded(selectedSubreddit));
		}
	}

	handleChange = nextSubreddit => {
		this.props.dispatch(selectSubreddit(nextSubreddit));
	};

	handleRefreshClick = e => {
		e.preventDefault();

		const {dispatch, selectedSubreddit} = this.props;
		// why invalidate?
		dispatch(invalidateSubreddit(selectedSubreddit));
		dispatch(fetchPostsIfNeeded(selectedSubreddit));
	};

	render() {
		const props = {...this.props, handleRefreshClick: this.handleRefreshClick, handleChange: this.handleChange};
		return <App {...props} />
	}
}

const mapStateToProps = state => {
	const {selectedSubreddit, postsBySubreddit} = state;
	const {isFetching, lastUpdated, items: posts} = postsBySubreddit[selectedSubreddit] || {
		isFetching: true,
		items: []
	};
	return {
		selectedSubreddit,
		posts,
		isFetching,
		lastUpdated
	}
};


export default connect(mapStateToProps)(RedditApp)