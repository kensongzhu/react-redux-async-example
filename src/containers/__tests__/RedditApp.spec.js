import React from "react";
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";
import RedditApp from "../RedditApp";
import {shallow} from "enzyme";

const setup = (initialState = {}, strategy = shallow) => {
    const store = configureStore([thunk])(initialState);
    const wrapper = strategy(<RedditApp store={store}/>);
    return {
        store,
        wrapper
    }
};


describe('Reddit App', () => {

    afterEach(() => {
        fetch.resetMocks();
    });

    test('should render without crashing', () => {
        fetch.once(JSON.stringify({
            data: {
                children: [
                    {data: {title: 'subreddit 1'}},
                    {data: {title: 'subreddit 2'}}
                ]
            }
        }));
        const {wrapper} = setup({selectedSubreddit: 'frontend', postsBySubreddit: {}});
        expect(wrapper).toMatchSnapshot();
    });
});
