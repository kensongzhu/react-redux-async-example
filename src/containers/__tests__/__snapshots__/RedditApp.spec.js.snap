// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reddit App should render without crashing 1`] = `
<ContextProvider
  value={null}
>
  <RedditApp
    dispatch={[Function]}
    isFetching={true}
    posts={Array []}
    selectedSubreddit="frontend"
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;
